File: python/solutions/day_02.py
FILE_PATH = "./../../csharp/InputFiles/02.txt"
TEST_FILE_PATH = "./../../csharp/InputFiles/02TestFile.txt"

def day_02(use_test_files = False):
    file_path = get_input_file(use_test_files)
    part_1(file_path)
    part_2(file_path)

def part_1(file_path):
    result = open_input_file(file_path)
    print(f'Day 02 Part 1 - {result}')

def part_2(file_path):
    result = open_input_file(file_path, True)
    print(f'Day 02 Part 2 - {result}')

def open_input_file(file_path, part_two = False):
    safe_count = 0
    with open(file_path, 'r') as file:
        for line in file:
            levels = line.split()
            if is_safe(levels, -1):
                safe_count += 1
            else:
                if part_two:
                    for i in range(len(levels)):
                        if is_safe(levels, i):
                            safe_count += 1
                            break

    return safe_count

def is_safe(arr, ignore_ordinal):
    tolerance = 3
    arr = [arr[i] for i in range(len(arr)) if i != ignore_ordinal]
    ascending = all(1 <= int(arr[i]) - int(arr[i + 1]) <= tolerance for i in range(len(arr) - 1))
    descending = all(-1 >= int(arr[i]) - int(arr[i + 1]) >= tolerance * -1 for i in range(len(arr) - 1))
    return ascending or descending

def get_input_file(use_test_files = False):
    return TEST_FILE_PATH if use_test_files else FILE_PATH

File: csharp/Solutions/Day02.cs
using AdventOfCode.InputFiles;

namespace AdventOfCode.Solutions
{
    public static class Day02
    {
        private const int Day = 2;

        public static class Part1
        {
            public static double GetAnswer()
            {
                string[] inputData = InputCache.Get(Day);
                return Shared.Result(inputData);
            }
        }

        public static class Part2
        {
            public static double GetAnswer()
            {
                string[] inputData = InputCache.Get(Day);
                return Shared.Result(inputData, true);
            }
        }

        private class Shared
        {
            public static double Result(string[] set, bool partTwo = false)
            {
                var result = 0d;
                for (int i = 0; i < set.Length; i++)
                {
                    if (set[i].Trim() == string.Empty)
                    {
                        continue;
                    }

                    var levels = set[i].Split(' ');
                    if (IsSafe(levels, -1))
                    {
                        result += 1;
                    }
                    else if (partTwo)
                    {
                        for (int e = 0; e < levels.Length; e++)
                        {
                            if (IsSafe(levels, e))
                            {
                                result += 1;
                                break;
                            }
                        }
                    }
                }
                return result;
            }

            public static bool IsSafe(string[] set, int ignoreOrdinal)
            {
                const int Tolerance = 3;
                var newSet = set.Where((value, index) => index != ignoreOrdinal).ToArray();
                bool ascending = true;
                bool descending = true;
                for (int i = 0; i < newSet.Length - 1; i++)
                {
                    var diff = Convert.ToInt32(newSet[i]) - Convert.ToInt32(newSet[i + 1]);
                    if (diff < 1 || diff > Tolerance)
                    {
                        ascending = false;
                        break;
                    }
                }
                for (int i = 0; i < newSet.Length - 1; i++)
                {
                    var diff = Convert.ToInt32(newSet[i]) - Convert.ToInt32(newSet[i + 1]);
                    if (diff > -1 || diff < Tolerance * -1)
                    {
                        descending = false;
                        break;
                    }
                }
                return ascending || descending;
            }
        }
    }
}
